sdk/dynamicconfig/metric/controller/push/push.go

```diff
@@ -0,0 +1,196 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package push
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	sdk "go.opentelemetry.io/contrib/sdk/dynamicconfig/metric"
+
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/metric/controller/notify"
+	nbasic "go.opentelemetry.io/contrib/sdk/dynamicconfig/metric/controller/notify/basic"
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/metric"
+	"go.opentelemetry.io/otel/api/metric/registry"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	controllerTime "go.opentelemetry.io/otel/sdk/metric/controller/time"
+	"go.opentelemetry.io/otel/sdk/metric/processor/basic"
+)
+
+const fallbackPeriod = 10 * time.Minute
+
+// Controller organizes a periodic push of metric data.
+type Controller struct {
+	lock        sync.Mutex
+	accumulator *sdk.Accumulator
+	provider    *registry.Provider
+	processor   *basic.Processor
+	exporter    export.Exporter
+	lastPeriod  time.Duration
+	quit        chan struct{}
+	done        chan struct{}
+	isRunning   bool
+	timeout     time.Duration
+	clock       controllerTime.Clock
+	ticker      controllerTime.Ticker
+	notifier    notify.Notifier
+	mch         notify.MonitorChannel
+	matcher     *PeriodMatcher
+}
+
+// New constructs a Controller, an implementation of metric.Provider,
+// using the provided exporter and options to configure an SDK with
+// periodic collection.
+func New(selector export.AggregatorSelector, exporter export.Exporter, configHost string, opts ...Option) *Controller {
+	c := &Config{}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
+	if c.Timeout == 0 {
+		c.Timeout = fallbackPeriod
+	}
+
+	processor := basic.New(selector, exporter)
+	impl := sdk.NewAccumulator(
+		processor,
+		sdk.WithResource(c.Resource),
+	)
+
+	notifier := nbasic.NewNotifier(configHost, c.Resource)
+	mch := notify.MonitorChannel{
+		Data: make(chan *notify.MetricConfig),
+		Err:  make(chan error),
+		Quit: make(chan struct{}),
+	}
+
+	return &Controller{
+		provider:    registry.NewProvider(impl),
+		accumulator: impl,
+		processor:   processor,
+		exporter:    exporter,
+		quit:        make(chan struct{}),
+		timeout:     c.Timeout,
+		clock:       controllerTime.RealClock{},
+		notifier:    notifier,
+		mch:         mch,
+		matcher:     &PeriodMatcher{},
+	}
+}
+
+// Provider returns a metric.Provider instance for this controller.
+func (c *Controller) Provider() metric.Provider {
+	return c.provider
+}
+
+// Start begins a ticker that periodically collects and exports
+// metrics with the configured interval.
+func (c *Controller) Start() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.isRunning {
+		return
+	}
+
+	c.isRunning = true
+	c.matcher.MarkStart(c.clock.Now())
+	go c.notifier.MonitorChanges(c.mch)
```

Comment:
Does it break the encapsulation of a class if we create a goroutine for notifier's internal method? Would it be better to let the notifier (or rather monitor) creates its monitoring goroutine internally?

Reply:
Great point, modifying logic to preserve encapsulation

Reply:
Been wrestling with a couple ideas for improving encapsulation and running into two main issues:
* moving the channel logic internally into the notifier necessitates a second layer of channels in order to allow the push controller to select on blocking operations. It seems a little sloppy, complicates the code, and may incur a performance hit
* the notifier interface requires 4 methods (start, stop, get error, get data) rather than just 1 (monitor changes), which makes the interface a little less easy to use

For the existing solution, since `MonitorChanges()` is an exported method, it isn't solidly part of the internal logic of the notifier. In this vein, `MonitorChannel` is more of a communication medium between the notifier and pusher, rather than an operation that needs to be encapsulated. It would be possible to move the `go` call inside notifier instead of the pusher,  which I've gone ahead and done.